specs/deneb/beacon-chain.md

```diff
@@ -224,7 +217,11 @@ def process_execution_payload(state: BeaconState, payload: ExecutionPayload, exe
     # Verify timestamp
     assert payload.timestamp == compute_timestamp_at_slot(state, state.slot)
     # Verify the execution payload is valid
-    assert execution_engine.notify_new_payload(payload)
+    # [Modified in Deneb]
+    versioned_hashes = [kzg_commitment_to_versioned_hash(commitment) for commitment in body.blob_kzg_commitments]
+    assert execution_engine.notify_new_payload(
```

Comment:
another strategy is to keep `notify_new_payload()` as the same and to insert another function `execution_engine.verify_versioned_hashes(versioned_hashes, payload.transactions)`. This would be divergent from the engine API but is very explicit.

In fact if we went that path, I would also break out the verification of the `block_hash` -- e.g.

```python
assert execution_engine.is_valid_block_hash(payload)
assert execution_engine.is_valid_versioned_hashes(versioned_hashes, payload.transactions)
assert execution_engine.notify_new_payload(payload)
```

Although I like how explicit this is, it is arguably not worth the complexity here in changing the abstraction (especially given the engine api will/should remain stable as one call)

Reply:
cc: @mkalinin

Reply:
I really like this strategy as making these checks explicit brings more clarity to the spec. Can't see anything that could be broken by doing it this way, I was thinking about optimistic sync tests but they shouldn't be affected

Reply:
But in client implementations, the validation will still use the one Engine API call response, correct? Aren't different API abstractions make it more difficult to test? e.g., we already yield `execution_valid` in test vectors, if we want to add what's proposed in https://github.com/ethereum/consensus-specs/pull/3359#discussion_r1197989374 + add `execution_engine.is_valid_versioned_hashes()` helper in specs, we have to yield an `is_valid_versioned_hashes` field in test vectors.

I'd like to confirm if clients indeed prefer parsing this themselves. 🤔

/cc @terencechain

Reply:
my view: we would just like it as a single call with length check before the call (check should come before versioned hashes computation but suggested here for the conversation context)
```suggestion
    asset body.blob_kzg_commitments <= MAX_BLOBS_PER_BLOCK
    assert execution_engine.notify_new_payload(
```

Reply:
I think separating `is_valid_versioned_hashes` with `notify_new_payload` makes sense only if there are cases that we want to call one without the other. If we always call both, I think it makes more sense to do a single call.

If we think that the name is not explicit enough, we can just change the name of `notify_new_payload` to be something like `notify_new_block`?

Reply:
Given this is a *specification* to show complete state transition validation and mutation logic and *not* an implementation guide that knows or cares about the engine api (Engine API is just one architecture of a valid full client), I very much like explicitly calling out the various groupings of validations instead of just hiding it behind one.

That said, at this point in the process, if it is going to cause too much additional overhead (e.g. with changing test formats), I'm not going to push too hard for it.

Reply:
@g11tech @ppopth

The extra length check is unnecessary since the SSZ type has defined `blob_kzg_commitments: List[KZGCommitment, MAX_BLOBS_PER_BLOCK]`.

Reply:
@djrtwo @mkalinin

I agree the explicitness is good, but CL is just the caller and has no control over executing the validation. I support @ppopth's proposal of renaming the API name or creating a new API to replace the old one in Deneb.

Another way is to make it clear in the abstract API:

```python
def notify_new_payload_and_verify_versioned_hashes(self: ExecutionEngine,
                                                   new_payload_request: NewPayloadRequest) -> bool:
    """
    Return ``True`` if and only if ``new_payload_request`` is valid with respect to ``self.execution_state``.
    """
    assert self.is_valid_block_hash(new_payload_request.execution_payload)
    assert self.is_valid_versioned_hashes(new_payload_request)
    assert self.notify_new_payload(new_payload_request.execution_payload)
```

And we stick to using the single `execution_valid` return value of `notify_new_payload_and_verify_versioned_hashes` in testing.

Note: we override this API in `setup.py`, so the content doesn't matter.

Reply:
> The extra length check is unnecessary since the SSZ container has defined field `blob_kzg_commitments: List[KZGCommitment, MAX_BLOBS_PER_BLOCK]`.

It will be changed in this PR https://github.com/ethereum/consensus-specs/pull/3338

Reply:
Ah...! but this PR will get merged first, so #3338 will have to address it. 😅